### ISSUE TITLE
"SdkClientException: Channel was closed before it could be written to." when running requests to CodeBuild in parallel

### ISSUE THREAD

**Issue by derwasp** [aws/aws-sdk-java-v2#1615]
After the update from `2.5.23` to `2.10.54` (actually to any of the versions between 2.10.49-2.10.54) parallel requests to CodeBuild started failing with `SdkClientException: Channel was closed before it could be written to.`

## Expected Behavior
Stuff doesn't crash

## Current Behavior
It does crash

## Possible Solution
No idea

## Steps to Reproduce (for bugs)
Doesn't always reproduce, so I can't provide you with a repro.
I tried doing it with a [demo project](https://github.com/albumprinter/junit-synnefo-demo) that uses our tool, but neither it nor our real test solution fail on my laptop after ~10 runs.

## Context
We are using CodeBuild to run out integration tests in parallel using an inhouse tool we've built using AWS Java SDK v2.
The home made tool can be found here, if you want to see at what it does:
https://github.com/albumprinter/junit-synnefo
What we do with the SDK is we create ~200 CodeBuild jobs in batches of 25 each 2500ms. The tool itself is running inside a CodeBuild job as well.

So it's something like this:
```
                    / [CodeBuild slave job 1]
[CodeBuild main job]  -- [CodeBuild slave job 2]
                    \  [CodeBuild slave job 3]
```
So our tool creates these slave jobs and just watches over them. Once they are done, we collect the artifacts for reports.

Previously we had v `2.5.23` and it still works just fine.

We've updated a bunch of dependencies a few days ago and started experiencing issues with the SDK calls to CodeBuild when we want to start a job.

This is the example exception:

```
software.amazon.awssdk.core.exception.SdkClientException: Channel was closed before it could be written to.
```
And a more detailed stack trace:

```
    software.amazon.awssdk.core.exception.SdkClientException: Channel was closed before it could be written to.
        at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:97)
        at software.amazon.awssdk.core.internal.util.ThrowableUtils.asSdkException(ThrowableUtils.java:98)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryIfNeeded(AsyncRetryableStage.java:126)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.lambda$execute$0(AsyncRetryableStage.java:108)
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
        at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$WrappedErrorForwardingResponseHandler.onError(MakeAsyncHttpRequestStage.java:124)
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.handleFailure(NettyRequestExecutor.java:267)
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.tryConfigurePipeline(NettyRequestExecutor.java:188)
        at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.makeRequestListener(NettyRequestExecutor.java:137)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551)
        at io.netty.util.concurrent.DefaultPromise.access$200(DefaultPromise.java:35)
        at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:501)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518)
        at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1050)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.lang.Thread.run(Thread.java:748)
    Caused by: java.io.IOException: Channel was closed before it could be written to.
        ... 12 more
```

Sometimes we also get
```
java.io.IOException: Server failed to send complete response
```
But I don't have a stacktrace for it.

For now we just [rolled back](https://github.com/albumprinter/junit-synnefo/pull/84/files), but we would like to keep our applications up to date.

Another thing is that we are running this in two different AWS accounts and for one account if fails more often than for another account.
I wasn't able to reproduce this while running it from a laptop on either Windows or MacOS. It just happens in CodeBuild

## Your Environment

* AWS Java SDK version used: `2.10.54`
* JDK version used: jdk8 from [CodeBuild standard-2.0](https://github.com/aws/aws-codebuild-docker-images/tree/master/ubuntu/standard/2.0)
* Operating System and version: [CodeBuild standard-2.0](https://github.com/aws/aws-codebuild-docker-images/tree/master/ubuntu/standard/2.0)

**Comment by debora-ito** [comment]
Hi @derwasp, thank you for reaching out. We are aware of the problem and currently investigating.

**Comment by zoewangg** [comment]
Hi @derwasp I noticed you are using S3 client as well, have you experienced issue with s3 client? or just codebuild client?

Looking at your code, it seems retry policy has been overridden for codebuild client and only service exception is retried. 

`SdkClientException: Channel was closed before it could be written to` indicates a soon-to-be-stale connection gets reused before it closes and it is thrown as `IOException`. Retry should work in this case if you add `IOException` in the retry condition.

**Comment by debora-ito** [comment]
Closing this due to inactivity. Feel free to reopen if you have more questions.

**Comment by Bennett-Lynch** [comment]
Hi @derwasp. A change was released today as part of version `2.17.101` that should help significantly reduce the occurrence of `Channel was closed before it could be written to` errors. If you are still experiencing this issue, please try the latest version and let us know if you see improvement. Change details can be found in https://github.com/aws/aws-sdk-java-v2/pull/2883.
